primary/src/block_synchronizer/mod.rs

```diff
@@ -333,6 +382,145 @@ impl BlockSynchronizer {
+        if self.sync_range_state.running {
+            return None;
+        }
+
+        // Initialize range sync state.
+        let (sender, receiver) = channel(self.committee.size());
+        assert!(self.sync_range_state.responded_peers.is_empty());
+        self.sync_range_state.running = true;
+        self.sync_range_state.result_sender = Some(respond_to);
+        self.sync_range_state.item_sender = Some(sender);
+
+        // Broadcast range sync request.
+        let last_round = self.certificate_store.last_round_number();
+        // NOTE: Assuming locally missing certificates in existing rounds are negligible issues,
+        // range sync starts from the next round where there is no certificate stored locally.
+        // We can switch to a more fine grained per-certificate-author range sync if necessary.
+        // Start sync from round 0, if there is no certificate in store.
+        let range_start = if let Some(r) = last_round { r + 1 } else { 0 };
+        let message = PrimaryMessage::CertificatesRangeRequest {
+            range_start,
+            max_rounds: CERTIFICATES_RANGE_REQUEST_MAX_ROUNDS,
+            requestor: self.name.clone(),
+        };
+        self.broadcast_batch_request(message.clone()).await;
+
+        // Responses are not await here to avoid blocking the BlockSynchronizer loop.
+        Some(
+            Self::wait_for_range_sync_responses(
+                receiver,
+                self.committee.clone(),
+                self.range_synchronize_timeout,
+            )
+            .boxed(),
+        )
+    }
+
+    async fn handle_digest_response(&mut self, certificate_digests: CertificateDigestsResponse) {
+        let CertificateDigestsResponse {
+            certificate_ids,
+            from,
+        } = certificate_digests;
+        if !self.sync_range_state.running {
+            trace!("dropping range sync request since range sync is active");
+            return;
+        }
+        if from == self.name {
+            trace!("dropping range sync request from the same authority");
+            return;
+        }
+        if self.committee.primary(&from).is_err() {
+            trace!("dropping range sync request from an origin that is not in the committe");
+            return;
+        }
+        let state = &mut self.sync_range_state;
+        if !state.responded_peers.insert(from.clone()) {
+            trace!(
+                "dropping range sync request from a peer we already heard from: {:?}",
+                from.clone()
+            );
+            return;
+        }
+        if certificate_ids.len() > CERTIFICATES_RANGE_REQUEST_MAX_ROUNDS as usize {
+            // Skip response containing too many rounds.
+            trace!(
+                "dropping range sync request containing too many rounds: {:?} > {:?}",
+                certificate_ids.len(),
+                CERTIFICATES_RANGE_REQUEST_MAX_ROUNDS
+            );
+            return;
+        }
```

Comment:
It would be great to print the origin of this `CertificateDigestsResponse` (`from`).